### PR TITLE
Automatically update the triage owner on Bugzilla

### LINKED ISSUE #1491: Round robin should also update triage owner on Bugzilla
This way the triage owner rotation is reflected in Bugzilla too

### PR DESCRIPTION
**Pull request by suhaibmujahid**
Resolves #1491 

Blocked by:
- https://github.com/mozilla/libmozdata/pull/201

## Dry-run

```
2022-07-01 00:21:44,468 - INFO - Run tool triage_owner_rotations.py
2022-07-01 00:22:00,730 - INFO - Tool triage_owner_rotations.py has finished.
```
<p>The following triage owners will be automatically assigned:
    <table style="border:1px solid black;border-collapse:collapse;" border="1">
    <thead>
    <tr>
    <th>Component</th><th>Old Triage Owner</th><th>New Triage Owner</th>
    </tr>
    </thead>
    <tbody>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::CSS Parsing and Computation
    </td>
    <td>
    emilio@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::CSS Transitions and Animations
    </td>
    <td>
    boris.chiou@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DMD
    </td>
    <td>
    continuation@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Bindings (WebIDL)
    </td>
    <td>
    echen@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Content Processes
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Copy & Paste and Drag & Drop
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Core & HTML
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: CSS Object Model
    </td>
    <td>
    emilio@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Editor
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Events
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: File
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Forms
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Geolocation
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: HTML Parser
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Navigation
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: postMessage
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Push Notifications
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Selection
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Serializers
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Service Workers
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: UI Events & Focus Handling
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Web Payments
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::DOM: Window and Location
    </td>
    <td>
    htsai@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::DOM: Workers
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::General
    </td>
    <td>
    overholt@xxxx.xxx
    </td>
    <td>
    sdetar@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Columns
    </td>
    <td>
    aethanyc@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout: Flexbox
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Floats
    </td>
    <td>
    aethanyc@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout: Form Controls
    </td>
    <td>
    emilio@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Generated Content, Lists, and Counters
    </td>
    <td>
    emilio@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout: Grid
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Positioned
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout: Ruby
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Scrolling and Overflow
    </td>
    <td>
    hikezoe.birchill@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Layout: Tables
    </td>
    <td>
    dholbert@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Layout: Text and Fonts
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Print Preview
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Printing: Output
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Printing: Setup
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Storage: Cache API
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Storage: IndexedDB
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Storage: localStorage & sessionStorage
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::Storage: Quota Manager
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Core::Storage: StorageManager
    </td>
    <td>
    jstutte@xxxx.xxx
    </td>
    <td>
    bugs@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Core::SVG
    </td>
    <td>
    jwatt@xxxx.xxx
    </td>
    <td>
    jfkthame@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Firefox::Installer
    </td>
    <td>
    bytesized@xxxx.xxx
    </td>
    <td>
    ksteuber@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Firefox::Shell Integration
    </td>
    <td>
    bytesized@xxxx.xxx
    </td>
    <td>
    ksteuber@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Firefox::Sync
    </td>
    <td>
    markh@xxxx.xxx
    </td>
    <td>
    bdeankawamura@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Testing::AWSY
    </td>
    <td>
    dave.hunt@xxxx.xxx
    </td>
    <td>
    gmierz2@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Testing::Performance
    </td>
    <td>
    dave.hunt@xxxx.xxx
    </td>
    <td>
    gmierz2@xxxx.xxx
    </td>
    </tr>
    <tr >
    <td>
    Toolkit::about:memory
    </td>
    <td>
    continuation@xxxx.xxx
    </td>
    <td>
    peterv@xxxx.xxx
    </td>
    </tr>
    <tr bgcolor="#E0E0E0">
    <td>
    Toolkit::Application Update
    </td>
    <td>
    bytesized@xxxx.xxx
    </td>
    <td>
    ksteuber@xxxx.xxx
    </td>
    </tr>
    </tbody>
    </table>
    </p>

## Checklist

- [x] Type annotations added to new functions
- [x] Docs added to functions touched in main classes
- [x] Dry-run produced the expected results

### CODE REVIEW

File: auto_nag/scripts/triage_owner_rotations.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    new_triager.component,
+                    change,
+                )
+            else:
+                try:
+                    BugzillaComponent(
+                        new_triager.product,
+                        new_triager.component,
+                    ).put(change)
+                except HTTPError as err:
```
Comment (suhaibmujahid) [inline auto_nag/scripts/triage_owner_rotations.py:45]: Maybe we should implement a retry in `libmozdata` for `BugzillaComponent`

Comment (marco-c) [reply auto_nag/scripts/triage_owner_rotations.py:45]: Or use a library like `tenacity` here, so we decouple the retrying from the implementation of the API calls.

Comment (suhaibmujahid) [reply auto_nag/scripts/triage_owner_rotations.py:45]: I like this option more (implement the retry here, not in the API). I initially suggested the retry in `libmozdata` for consistency:

https://github.com/mozilla/libmozdata/blob/c092351450dae908c4f7184bebb492d304910035/libmozdata/bugzilla.py#L101

**Comment by sylvestre** [comment]
cc @globau

File: auto_nag/scripts/triage_owner_rotations.py (hunk 1 above)
Comment (suhaibmujahid) [reply auto_nag/scripts/triage_owner_rotations.py]: Done in 7913ab960c9dacb7310d7781428a3e45fa871006

File: auto_nag/scripts/triage_owner_rotations.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        self, new_triage_owners: List[TriageOwner]
+    ) -> Set[ComponentName]:
+        failures = set()
+        for new_triager in new_triage_owners:
+            if self.dryrun or self.test_mode:
+                logger.info(
+                    "The triage owner for '%s' will be: '%s'",
+                    new_triager.component,
+                    new_triager.bugzilla_email,
+                )
```
Comment (marco-c) [inline auto_nag/scripts/triage_owner_rotations.py:61]: Nit: maybe always log this, so remove the if and replace the else with `is not self.dryrun and not self.test_mode`

Comment (marco-c) [reply auto_nag/scripts/triage_owner_rotations.py:61]: As a side note: what is "test mode"? I had never heard of it :P

**Review by marco-c — DISMISSED** [review]

File: auto_nag/scripts/triage_owner_rotations.py (hunk 2 above)
Comment (suhaibmujahid) [reply auto_nag/scripts/triage_owner_rotations.py]: > As a side note: what is "test mode"? I had never heard of it :P

Me too, I just matched `BzCleaner` behaviour; it seems a configuration that can be set in the `tools.json` file:

https://github.com/mozilla/relman-auto-nag/blob/179462fd873a93aaca104196479483f2e728279b/auto_nag/bzcleaner.py#L576

https://github.com/mozilla/relman-auto-nag/blob/aa8da65e7d6b4364c0994644180e9195daef0905/auto_nag/scripts/configs/tools.json#L43




> Nit: maybe always log this, so remove the if and replace the else with `is not self.dryrun and not self.test_mode`

Done in d3756d6184cbdcaeb785ca6bd16bfcaa72e1df2b

File: auto_nag/scripts/triage_owner_rotations.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+from auto_nag import logger
+from auto_nag.bzcleaner import BzCleaner
+from auto_nag.component_triagers import ComponentName, ComponentTriagers, TriageOwner
+
+
+class TriageOwnerRotations(BzCleaner):
+    def __init__(
+        self,
+        excluded_components: List[str] = [],
```
Comment (marco-c) [inline auto_nag/scripts/triage_owner_rotations.py:26]: Nit: given that this is empty, perhaps we can remove it and hope we'll never need it.

**Review by marco-c — APPROVED** [review]

**Pull request merged**